### PR TITLE
Fix bad LOD selection when Camera in Mesh AABB

### PR DESCRIPTION
**Pull request by 0010200303**
Fixed an issue where a bad LOD would be selected although the Camera being inside the Meshes AABB.

### CODE REVIEW

**Comment by clayjohn** [comment]
Related to https://github.com/godotengine/godot/pull/79191

**Comment by 0010200303** [comment]
Yeah it fixes the MRP in #67890 but i don't really know if its a better solution than #79191. I primarily tried to fix an issue when the Camera is inside a really big MultiMeshInstance.

**Comment by clayjohn** [comment]
CC @fire Does this PR fix your use-case?

**Comment by fire** [comment]
I'll try to review later today. @lyuma too. I'll run my test case.

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp [hunk 1]
```diff
@@ -944,25 +944,33 @@ void RenderForwardClustered::_fill_render_list(RenderListType p_render_list, con
...
-				} else if (distance_max <= 0.0) {
-					distance = -distance_max;
+				// Check if camera is NOT inside the mesh AABB.
+				Vector3 camera_position = p_render_data->scene_data->cam_transform.origin;
+				Vector3 aabb_start = inst->transformed_aabb.get_position();
+				Vector3 aabb_end = inst->transformed_aabb.get_end();
+
+				if ((aabb_start.x <= camera_position.x && camera_position.x <= aabb_end.x &&
+							aabb_start.y <= camera_position.y && camera_position.y <= aabb_end.y &&
+							aabb_start.z <= camera_position.z && camera_position.z <= aabb_end.z) == false) {
```
Comment (lyuma) [inline servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:956]:
```suggestion
				if (!inst->transformed_aabb.has_point(camera_position)) {
```

**Review by lyuma — COMMENTED** [review]
Looks reasonable to me: this should at a minimum prevent changing LOD while inside the bounding box which is one of the bugs that you currently get on a large map.

Note that there was another PR open which was attempting to solve the case of inside LOD in a different way by scaling it. See #79191 - I don't quite understand the mechanics of how it works but I just wanted to mention it.

The rest of the existing code that got moved inside the if statement still seems highly suspect, but that can be solved separately. In particular, if you imagine a FOV of 179.9 (the extreme case), looking parallel to the nearest side of the AABB, `p_render_data->scene_data->cam_transform.basis.get_column(Vector3::AXIS_Z)` will return a vector roughly perpendicular to the nearest wall which could mean `inst->transformed_aabb.get_support();` is going to return a point far away. I think the LOD system still has weird issues depending on the angle you look at the bounding box, and I suspect its' related to the use of the forward vector on the camera rather than the vector from the camera to the AABB.

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp (hunk 1 above)
Comment (lyuma) [reply servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:956]: I don't see any difference between the if statement you wrote and the `AABB::has_point()` function. Is there something I'm missing here?

Comment (0010200303) [reply servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:956]: Yes you are right, it is the same. At the time i didn't knew AABB had the has_point function.

**Comment by fire** [comment]
Please use `AABB::has_point()`.

It doesn't solve my use case of giving a heuristic guess for decimation of avatars given they're inside the aabb.

The consensus is that picking distance 0 (which we've discussion is not the original mesh) for when the camera is inside the AABB.

We don't have consensus for the heuristic guess.

**Comment by clayjohn** [comment]
Looks great! The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable. 

If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

**Review by clayjohn — APPROVED** [review]
Looks great! Thank you

**Comment by YuriSizov** [comment]
Thanks! And congrats on your first merged Godot pull request!

**Pull request merged**